### PR TITLE
bpo-37958: Adding get_profile_dict to pstats

### PR DESCRIPTION
**Pull request by Olshansk**
pstats is really useful or profiling and printing the output of the execution of some block of code, but I've found on multiple occasions when I'd like to access this output directly in an easily usable dictionary on which I can further analyze or manipulate.

The proposal is to add a function called get_profile_dict inside of pstats that'll automatically return this data the data in an easily accessible dict.

The output of the following script:

```
import cProfile, pstats
import pprint
from pstats import func_std_string, f8

def fib(n):
    if n == 0:
        return 0
    if n == 1:
        return 1
    return fib(n-1) + fib(n-2)

pr = cProfile.Profile()
pr.enable()
fib(5)
pr.create_stats()

ps = pstats.Stats(pr).sort_stats('tottime', 'cumtime')

def get_profile_dict(self, keys_filter=None):
    """
        Returns a dict where the key is a function name and the value is a dict
        with the following keys:
            - ncalls
            - tottime
            - percall_tottime
            - cumtime
            - percall_cumtime
            - file_name
            - line_number

        keys_filter can be optionally set to limit the key-value pairs in the
        retrieved dict.
    """
    pstats_dict = {}
    func_list = self.fcn_list[:] if self.fcn_list else list(self.stats.keys())

    if not func_list:
        return pstats_dict

    pstats_dict["total_tt"] = float(f8(self.total_tt))
    for func in func_list:
        cc, nc, tt, ct, callers = self.stats[func]
        file, line, func_name = func
        ncalls = str(nc) if nc == cc else (str(nc) + '/' + str(cc))
        tottime = float(f8(tt))
        percall_tottime = -1 if nc == 0 else float(f8(tt/nc))
        cumtime = float(f8(ct))
        percall_cumtime = -1 if cc == 0 else float(f8(ct/cc))
        func_dict = {
            "ncalls": ncalls,
            "tottime": tottime, # time spent in this function alone
            "percall_tottime": percall_tottime,
            "cumtime": cumtime, # time spent in the function plus all functions that this function called,
            "percall_cumtime": percall_cumtime,
            "file_name": file,
            "line_number": line
        }
        func_dict_filtered = func_dict if not keys_filter else { key: func_dict[key] for key in keys_filter }
        pstats_dict[func_name] = func_dict_filtered

    return pstats_dict

pp = pprint.PrettyPrinter(depth=6)
pp.pprint(get_profile_dict(ps))
```

will produce:

```
{"<method 'disable' of '_lsprof.Profiler' objects>": {'cumtime': 0.0,
                                                      'file_name': '~',
                                                      'line_number': 0,
                                                      'ncalls': '1',
                                                      'percall_cumtime': 0.0,
                                                      'percall_tottime': 0.0,
                                                      'tottime': 0.0},
 'create_stats': {'cumtime': 0.0,
                  'file_name': '/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/cProfile.py',
                  'line_number': 50,
                  'ncalls': '1',
                  'percall_cumtime': 0.0,
                  'percall_tottime': 0.0,
                  'tottime': 0.0},
 'fib': {'cumtime': 0.0,
         'file_name': 'get_profile_dict.py',
         'line_number': 5,
         'ncalls': '15/1',
         'percall_cumtime': 0.0,
         'percall_tottime': 0.0,
         'tottime': 0.0},
 'total_tt': 0.0}
 ```

 As an example, this can be used to generate a stacked column chart using various visualization tools which will assist in easily identifying program bottlenecks.

https://bugs.python.org/issue37958

Automerge-Triggered-By: @gpshead

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

**Comment by pablogsal** [comment]
Hi @Olshansk and thank you for opening a PR. Before we can review, could you:

* Open an issue on the Python issue tracker so we can discuss the feature (https://bugs.python.org/)
* Sign the CLA (check out @the-knights-who-say-ni comment).

**Comment by Olshansk** [comment]
@pablogsal I was wondering whom I should message to get some feedback about this idea?

**Review by gpshead — CHANGES_REQUESTED** [review]
This needs unittest coverage (see `Lib/test/test_pstats.py`).

File: Misc/NEWS.d/next/Library/2019-08-27-03-57-25.bpo-37958.lRORI3.rst [hunk 1]
```diff
@@ -0,0 +1,7 @@
+pstats is really useful or profiling and printing the output of the execution of some block of code, but I've found on multiple occasions when I'd like to access this output directly in an easily usable dictionary on which I can further analyze or manipulate.
```
Comment (gpshead) [inline Misc/NEWS.d/next/Library/2019-08-27-03-57-25.bpo-37958.lRORI3.rst]: A news entry should just state what is new, not be text selling it or proposing it.

Something mundane such as "Added the pstats.Stats.get_profile_dict() method to get the profile data as a dictionary."

File: Lib/pstats.py [hunk 2]
```diff
@@ -333,6 +333,45 @@ def eval_print_amount(self, sel, list, msg):
 
         return new_list, msg
 
+    def get_profile_dict(self, keys_filter=None):
+        """
+        Returns a dict where the key is a function name and the value is a
+        dict with the following keys:
```
Comment (gpshead) [inline Lib/pstats.py]: Is returning a dict with keys ideal?  That feels old style vs returning an object with attributes.  A dataclass perhaps?

File: Lib/pstats.py [hunk 3]
```diff
@@ -333,6 +333,45 @@ def eval_print_amount(self, sel, list, msg):
 
         return new_list, msg
 
+    def get_profile_dict(self, keys_filter=None):
```
Comment (gpshead) [inline Lib/pstats.py]: Document the new method in Doc/library/profile.rst, including a versionadded 3.9 tag.

File: Lib/pstats.py (hunk 2 above)
Comment (gpshead) [reply Lib/pstats.py]: related to this, is there a good reason to have the keys_filter feature?  (it wouldn't make as much sense on a dataclass, i guess filtered fields could be None?)

**Review by gpshead — CHANGES_REQUESTED** [review]

File: Misc/NEWS.d/next/Library/2019-08-27-03-57-25.bpo-37958.lRORI3.rst (hunk 1 above)
Comment (Olshansk) [reply Misc/NEWS.d/next/Library/2019-08-27-03-57-25.bpo-37958.lRORI3.rst]: Done.

File: Lib/pstats.py (hunk 2 above)
Comment (Olshansk) [reply Lib/pstats.py]: Thank you for the suggestion! I removed remove `keys_filter` to keep things clean and changed to using `dataclass`, which I didn't know about before but it's awesome.

File: Lib/pstats.py (hunk 3 above)
Comment (Olshansk) [reply Lib/pstats.py]: Done

**Comment by Olshansk** [comment]
I have made the requested changes; please review again.

Sorry for being MIA and taking so long to followup on this, but I've replied to all the comments and would appreciate if you take another look!

**Comment by Olshansk** [comment]
@gpshead Just wanted to follow up on this.

File: Lib/test/test_pstats.py [hunk 4]
```diff
@@ -75,6 +77,25 @@ def test_sort_starts_mix(self):
...
+        pr.enable()
+        pass1()
+        pass2()
+        pass3()
+        pr.create_stats()
+        ps = pstats.Stats(pr)
+
+        stats_profile = ps.get_stats_profile()
+        funcs_called = stats_profile.func_profiles.keys()
+        self.assertTrue(
```
Comment (gpshead) [inline Lib/test/test_pstats.py]: `assertIn({'pass1', ...}, set(funcs_called))` would generate a nicer failure message.

File: Lib/pstats.py [hunk 5]
```diff
@@ -333,6 +351,41 @@ def eval_print_amount(self, sel, list, msg):
...
 
+    def get_stats_profile(self):
+        """This method returns an instance of StatsProfile, which contains a mapping
+        of function names to instances of FunctionProfile. Each FunctionProfile
+        instance holds information related to the function's profile such as how
+        long the function took to run, how many times it was called, etc...
+        """
+        func_list = self.fcn_list[:] if self.fcn_list else list(self.stats.keys())
+        if not func_list:
+            return pstats_dict
```
Comment (gpshead) [inline Lib/pstats.py]: Where does this come from?

**Review by gpshead — CHANGES_REQUESTED** [review]

File: Lib/pstats.py (hunk 5 above)
Comment (Olshansk) [reply Lib/pstats.py]: Thanks for the catch, I forgot to update this!

I changed this to `StatsProfile(0, {})`. Not sure if we should go with this default value or raise an exception instead?

File: Lib/test/test_pstats.py (hunk 4 above)
Comment (Olshansk) [reply Lib/test/test_pstats.py]: assertIn (the way you have it above) didn't quite work to check if the first set is a complete subset of the second.

I modified it to the following:
```
        self.assertIn('pass1', funcs_called)
        self.assertIn('pass2', funcs_called)
        self.assertIn('pass3', funcs_called)
```

Let me know if this is what you were looking for.

**Comment by Olshansk** [comment]
I have made the requested changes; please review again.

File: Lib/pstats.py (hunk 5 above)
Comment (gpshead) [reply Lib/pstats.py]: what you've done looks reasonable, I could see it being done either way.

**Review by gpshead — APPROVED** [review]

**Comment by Olshansk** [comment]
This is awesome. Thanks a lot for your help @gpshead for helping me get this in!

File: Misc/NEWS.d/next/Library/2019-08-27-03-57-25.bpo-37958.lRORI3.rst [hunk 6]
```diff
@@ -0,0 +1,2 @@
+Added the pstats.Stats.get_profile_dict() method to return the profile
```
Comment (Olshansk) [inline Misc/NEWS.d/next/Library/2019-08-27-03-57-25.bpo-37958.lRORI3.rst:1]: @gpshead I just realized that I never updated the function name from from `get_profile_dict` to `get_stats_profile` after refactoring things from a dict to a dataclass. 

Should I submit another PR that updates this file?
